### PR TITLE
SONARPY-1470: ReachingDefinitionAnalysis should work with annotated assignments

### PR DESCRIPTION
**Pull request by joke1196**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/cfg/fixpoint/ReachingDefinitionsAnalysis.java [hunk 1]
```diff
@@ -186,26 +190,55 @@ private static Map<Symbol, Set<Expression>> join(Map<Symbol, Set<Expression>> pr
     return result;
   }
 
-  private void updateProgramState(Tree element, Map<Symbol, Set<Expression>> programState) {
+  private void updateProgramState(Tree element, Map<Symbol, Set<Expression>> out) {
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/cfg/fixpoint/ReachingDefinitionsAnalysis.java:193]: Looks good but I'd sugest following changes:
```java
private void updateProgramState(Tree element, Map<Symbol, Set<Expression>> out) {
    if (element.is(ASSIGNMENT_STMT)) {
      updateStateForAssignment((AssignmentStatement) element, out);
    }
    if (element.is(Tree.Kind.ANNOTATED_ASSIGNMENT)) {
      updateStateForAnnotatedAssignement((AnnotatedAssignment) element, out);
    }
  }

  private void updateStateForAssignment(AssignmentStatement element, Map<Symbol, Set<Expression>> programState) {
    List<Expression> lhsExpressions = element.lhsExpressions().stream()
      .flatMap(exprList -> exprList.expressions().stream())
      .collect(Collectors.toList());
    performUpdate(programState, lhsExpressions, element::assignedValue);
  }
  // Extend common logic from methods above
  private void updateStateForAnnotatedAssignement(AnnotatedAssignment element,
    Map<Symbol, Set<Expression>> programState) {
    List<Expression> lhsExpressions = List.of(element.variable());
    performUpdate(programState, lhsExpressions, element::assignedValue);
  }
  
  private void performUpdate(Map<Symbol, Set<Expression>> programState, List<Expression> lhsExpressions, Supplier<Expression> assignedValueSupplier) {
    getLhsExpression(lhsExpressions)
      .ifPresent(name -> Optional.of(name)
        .flatMap(TreeUtils::getSymbolFromTree)
        .ifPresent(symbol -> {
          var assignedValue = assignedValueSupplier.get();
          performUpdate(symbol, assignedValue, name, programState);
        })
      );
  }

  private void performUpdate(Symbol symbol, Expression assignedValue, Name lhsExpression,
    Map<Symbol, Set<Expression>> programState) {
    assignedNamesBySymbol.computeIfAbsent(symbol, s -> new HashSet<>()).add(lhsExpression);
    Set<Expression> assignedValues = new HashSet<>();
    assignedValues.add(assignedValue);
    // performing a strong update
    programState.put(symbol, assignedValues);
  }
  // You anyway expect to get optional Name in all the cases you call the method
  private static Optional<Name> getLhsExpression(List<Expression> lhsExpressions) {
    return Optional.of(lhsExpressions)
      .filter(l -> l.size() == 1)
      .stream()
      .flatMap(Collection::stream)
      .findFirst()
      .flatMap(TreeUtils.toOptionalInstanceOfMapper(Name.class));
  }
```

**Review by maksim-grebeniuk-sonarsource — CHANGES_REQUESTED** [review]
Looks good, just some logic could be extended as a common part. Also please make sure that formatting is done properly on code.

File: python-frontend/src/main/java/org/sonar/python/cfg/fixpoint/ReachingDefinitionsAnalysis.java (hunk 1 above)
Comment (joke1196) [reply python-frontend/src/main/java/org/sonar/python/cfg/fixpoint/ReachingDefinitionsAnalysis.java:197]: Thank you very much! Such a shame there are no tuples in Java by default.

**Review by maksim-grebeniuk-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**